### PR TITLE
Android run return

### LINKED ISSUE #1603: [Android] Run Return for Android
Currently run return is only available for Desktop operating systems. But at Embark Studios we require `run_return` for android as well.

### PR DESCRIPTION
**Pull request by VZout**
- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Fixes #1603

### CODE REVIEW

File: src/platform_impl/windows/event_loop/runner.rs [hunk 1]
```diff
@@ -250,7 +250,8 @@ impl<T> EventLoopRunner<T> {
         loop {
             // We do this instead of using a `while let` loop because if we use a `while let`
             // loop the reference returned `borrow_mut()` doesn't get dropped until the end
-            // of the loop's body and attempts to add events to the event buffer while in
```
Comment (filnet) [inline src/platform_impl/windows/event_loop/runner.rs:253]: seems unrelated

Comment (VZout) [reply src/platform_impl/windows/event_loop/runner.rs]: Good catch!

**Comment by VZout** [comment]
Opened the PR. CI appears to be failing due to a unrelated issue.

**Comment by francesca64** [comment]
Thanks for contributing again!

@dvc94ch are you fine with being pinged to review things like this? You're not listed on the [table](https://github.com/rust-windowing/winit/wiki/Testers-and-Contributors), but you seem to be the current expert. Let me know if I shouldn't pester you!

File: src/platform_impl/android/mod.rs [hunk 2]
```diff
@@ -44,6 +44,20 @@ fn poll(poll: Poll) -> Option<EventSource> {
...
+}
+
+macro_rules! call_event_handler {
+    ( $event_handler:expr, $window_target:expr, $cf:expr, $event:expr ) => {{
+        if $cf != ControlFlow::Exit {
+            $event_handler($event, $window_target, &mut $cf);
+        } else {
+            $event_handler($event, $window_target, &mut ControlFlow::Exit);
+        }
+    }};
```
Comment (dvc94ch) [inline src/platform_impl/android/mod.rs:60]: Can you explain what this macro is for? The if else block seems redundant

**Comment by dvc94ch** [comment]
@francesca64 it's fine, I'm not watching the repo, feel free to cc me on Android stuff.

File: src/platform_impl/android/mod.rs (hunk 2 above)
Comment (VZout) [reply src/platform_impl/android/mod.rs:60]: Its pretty much copied from the windows implementation.

if the user sets the control flow at `MainEventsCleared` but later changes the control flow to `wait` in the `RedrawRequested` event it won't exit anymore. This only allows you to change the control flow if the control flow is not exit.

This happens for example every single event with the `window_run_return` example which always sets the `*control_flow = ControlFlow::Wait;` back to wait even if the control flow is exit already. See here: 
https://github.com/rust-windowing/winit/blob/master/examples/window_run_return.rs#L31

Comment (dvc94ch) [reply src/platform_impl/android/mod.rs:60]: Oh I see, you're passing a mutable reference, my bad

File: src/platform_impl/android/mod.rs [hunk 3]
```diff
@@ -160,50 +203,66 @@ impl<T: 'static> EventLoop<T> {
...
                             self.window_target(),
-                            &mut cf,
+                            control_flow,
+                            event::Event::UserEvent(event)
                         );
                     }
                 }
-                None => {}
+                None => {
+                    control_flow = ControlFlow::Exit;
```
Comment (msiglreith) [inline src/platform_impl/android/mod.rs:215]: This causes my current application to restart in a loop.
Not sure what's the intention here as `first_event` is initialized as `None` on `EventLoop` creation which causes this match to always end up in the branch imo

Comment (VZout) [reply src/platform_impl/android/mod.rs:215]: Are you using `run_return` directly or `return`?

While the control flow is set to exit there the first time it is not necessarily permanent. There is a chance for the user to change it later to something else when `MainEventsCleared` or `RedrawRequested` is emitted. Than the first event is set to something other than none. Also the next time we enter the loop `first_event` won't (unless the poll returns None) be None anymore because it is assigned to something else down below.

Comment (VZout) [reply src/platform_impl/android/mod.rs:215]: Ah I see why this would be problematic when using `run`.

Comment (VZout) [reply src/platform_impl/android/mod.rs:215]: This works with the winit `window` example which uses `run` because the control flow is always set to ` *control_flow = ControlFlow::Wait;` by the user.

Comment (msiglreith) [reply src/platform_impl/android/mod.rs:215]: > Are you using run_return directly or return

`run`

> While the control flow is set to exit there the first time it is not necessarily permanent. There is a chance for the user to change it later to something else when MainEventsCleared or RedrawRequested is emitted.

That's not the case as far as I can see. In particular, that's caused by the way `call_event_handler` macro is implemented (https://github.com/VZout/winit/blob/6ce0fa2f09f94ecfad6ac3c37732c6c1f39f2401/src/platform_impl/android/mod.rs#L58). It creates a local variable which will throw away the value we set the control flow in our event handler.

> This works with the winit window example which uses run because the control flow is always set to *control_flow = ControlFlow::Wait; by the user.

Maybe, my phone works different, but I also overwrite it to `ControlFlow::Wait` at the beginning.

Comment (VZout) [reply src/platform_impl/android/mod.rs:215]: Hm your right. The call_event_handler would prevent that. I remember testing this with `run` and it working but maybe I messed something up.
